kbfsgit/runner.go

```diff
@@ -503,6 +514,196 @@ func (r *runner) processGogitStatus(
 	r.errput.Write([]byte("\n"))
 }
 
+// recursiveByteCount returns a sum of the size of all files under the
+// directory represented by `fs`, added to `totalSoFar`.  It also
+// returns the length of the last string it printed to `r.errput`.
+func (r *runner) recursiveByteCount(
+	ctx context.Context, fs billy.Filesystem, totalSoFar int64, toErase int) (
+	bytes int64, toEraseRet int, err error) {
+	fileInfos, err := fs.ReadDir("/")
+	if err != nil {
+		return 0, 0, err
+	}
+
+	for _, fi := range fileInfos {
+		if fi.IsDir() {
+			chrootFS, err := fs.Chroot(fi.Name())
+			if err != nil {
+				return 0, 0, err
+			}
+			totalSoFar, toErase, err = r.recursiveByteCount(
+				ctx, chrootFS, totalSoFar, toErase)
+			if err != nil {
+				return 0, 0, err
+			}
+		} else {
+			totalSoFar += fi.Size()
+			if r.progress {
+				// This function only runs if r.verbosity >= 1.
+				eraseStr := strings.Repeat("\b", toErase)
+				newStr := fmt.Sprintf("%d bytes... ", totalSoFar)
+				toErase = len(newStr)
+				r.errput.Write([]byte(eraseStr + newStr))
+			}
+		}
+	}
+	return totalSoFar, toErase, nil
+}
+
+// statusWriter is a simple io.Writer shim that logs to `r.errput` the
+// number of bytes written to `output`.
+type statusWriter struct {
+	r           *runner
+	output      io.Writer
+	soFar       int64
+	totalBytes  int64
+	nextToErase int
+}
+
+var _ io.Writer = (*statusWriter)(nil)
+
+func (sw *statusWriter) Write(p []byte) (n int, err error) {
+	n, err = sw.output.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	sw.soFar += int64(len(p))
+	eraseStr := strings.Repeat("\b", sw.nextToErase)
+	newStr := fmt.Sprintf("%d/%d bytes... ", sw.soFar, sw.totalBytes)
+	sw.r.errput.Write([]byte(eraseStr + newStr))
+	sw.nextToErase = len(newStr)
+	return n, nil
+}
+
+func (r *runner) copyFile(
+	ctx context.Context, fs billy.Filesystem, localFS billy.Filesystem,
```

Comment:
Perhaps name them `from` and `to` rather than `fs` and `localFS` in case we need the other direction?